server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldSyntheticWriterHelper.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper.flattened;
+
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A helper class that reconstructs the field including keys and values
+ * mapped by {@link FlattenedFieldMapper}. It looks at key/value pairs,
+ * parses them and uses the {@link XContentBuilder} to reconstruct the
+ * original object.
+ *
+ * Example: given the following set of sorted key value pairs read from
+ * doc values:
+ *
+ * "id": "AAAb12"
+ * "package.id": "102938"
+ * "root.asset_code": "abc"
+ * "root.from": "New York"
+ * "root.object.value": "10"
+ * "root.object.items": "102202929"
+ * "root.object.items": "290092911"
+ * "root.to": "Atlanta"
+ *
+ * it turns them into the corresponding object using {@link XContentBuilder}.
+ * If, for instance JSON is used, it generates the following after deserialization:
+ *
+ * `{
+ *     "id": "AAAb12",
+ *     "package": {
+ *         "id": "102938"
+ *     },
+ *     "root": {
+ *         "asset_code": "abc",
+ *         "from": "New York",
+ *         "object": {
+ *             "items": ["102202929", "290092911"],
+ *         },
+ *         "to": "Atlanta"
+ *     }
+ * }`
+ *
+ */
+class FlattenedFieldSyntheticWriterHelper {
+
+    private static class Prefix {
+        final List<String> prefix;
+
+        Prefix() {
+            this.prefix = Collections.emptyList();
+        }
+
+        Prefix(final String key) {
+            final String[] keyAsArray = key.split("\\.");
+            this.prefix = new ArrayList<>(Arrays.asList(keyAsArray).subList(0, keyAsArray.length - 1));
```

Comment:
Maybe creating the new array list can be avoided here? Something like: `List.of(keyAsArray).subList(0, keyAsArray.length - 1);`?